src/App.tsx

```diff
@@ -15,49 +15,105 @@ export const goodsFromServer = [
   'Garlic',
 ];
 
+enum SortType {
+  NONE,
+  ALPHABET,
+  LENGTH,
+}
+
+export function getReorderedGoods(
+  goods: string[],
+  sortType: SortType,
+  isReversed: boolean,
+) {
+  const visibleGoods = [...goods];
+
+  visibleGoods.sort((g1, g2) => {
+    switch (sortType) {
+      case SortType.ALPHABET:
+        return g1.localeCompare(g2);
+      case SortType.LENGTH:
+        return g1.length - g2.length;
+      default:
+        return 0;
+    }
+  });
+
+  return isReversed ? visibleGoods.reverse() : visibleGoods;
+}
+
 export const App: React.FC = () => {
+  const [isReversed, setIsReversed] = React.useState(false);
+  const [sortType, setSortType] = React.useState(SortType.NONE);
+
+  const sortByName = () => setSortType(SortType.ALPHABET);
+
+  const sortByLength = () => setSortType(SortType.LENGTH);
+
+  const reverse = () => setIsReversed(!isReversed);
+
+  const reset = () => {
+    setSortType(SortType.NONE);
+    setIsReversed(false);
+  };
+
   return (
     <div className="section content">
-      <div className="buttons">
+      <div className="buttons is-flex is-centered">
         <button
+          onClick={sortByName}
           type="button"
-          className="button is-info is-light"
+          className={`button is-outlined is-success is-rounded ${sortType !== SortType.ALPHABET && 'is-light'}`}
         >
           Sort alphabetically
         </button>
 
         <button
+          onClick={sortByLength}
           type="button"
-          className="button is-success is-light"
+          className={`button is-rounded is-danger is-outlined ${sortType !== SortType.LENGTH && 'is-light'}`}
         >
           Sort by length
         </button>
 
         <button
+          onClick={reverse}
           type="button"
-          className="button is-warning is-light"
+          className={`button is-rounded is-black is-outlined ${!isReversed && 'is-light'}`}
         >
           Reverse
         </button>
 
-        <button
-          type="button"
-          className="button is-danger is-light"
-        >
-          Reset
-        </button>
+        {(sortType || isReversed)
+          ? (
+            <button
+              type="button"
+              className="button is-rounded is-ghost is-outlined"
+              onClick={reset}
+            >
+              Reset
+            </button>
+          )
+          : ''}
+      </div>
+      <div className="is-flex is-justify-content-center">
+        <div className="has-text-centered">
+          {(getReorderedGoods(goodsFromServer, sortType, isReversed))
```

Comment:
you will call this bad boy every time a render happens
you can create a variable with these goods before the return, memoize it (useMemo with a dependency array, so it will be reordered only when `goodsFromServer, sortType, isReversed` change) and then use it inside of the return

it's not too big of a problem here, but if this value was not just inside div, but passed to another component, it would make an impact, because that component (and possibly its children would be rerendered every single time, regardless of whether `goodsFromServer, sortType, isReversed` changed)

Reply:
I'll need to do some research before implementing that, the rest was fixed